[clang] Refactor clang's keyword enable/disable mechanism to allow lldb to re-use it

dsandersllvm opened the pull request:
lldb's CPlusPlusNameParser is currently identifying keywords using it's own map implemented using clang/Basic/TokenKinds.def. However, it does not respect the language options so identifiers can incorrectly determined to be keywords when using languages in which they are not keywords.

Rather than implement the logic to enable/disable keywords in both projects it makes sense for lldb to use clang's implementation.

See #164284 for more information

dsandersllvm committed:
[clang] Refactor clang's keyword enable/disable mechanism to allow lldb to re-use it

lldb's CPlusPlusNameParser is currently identifying keywords using it's own
map implemented using clang/Basic/TokenKinds.def. However, it does not respect
the language options so identifiers can incorrectly determined to be keywords
when using languages in which they are not keywords.

Rather than implement the logic to enable/disable keywords in both projects it
makes sense for lldb to use clang's implementation.

See #164284 for more information

dsandersllvm committed:
fixup: formatting

Michael137 reviewed (approved):
LGTM (modulo two minor comments). Thanks for splitting this out! Could now even be unit-testable (not that you should do that as part of this PR though). Please wait a couple of days for the Clang maintainers to approve before merging

AaronBallman reviewed (approved):
LGTM modulo the requests from @Michael137

dsandersllvm committed:
fixup: Move getKeywordStatus's comment to the header

dsandersllvm committed:
fixup: Restore comment for TokenKey

This comment was originally on the namespace containing TokenKey and
KeywordStatus. I've only restored it to TokenKey because KeywordStatus
is used by the code that uses TokenKinds.def not TokenKinds.def itself.

dsandersllvm commented on clang/lib/Basic/IdentifierTable.cpp:
```diff
@@ -77,57 +77,6 @@ IdentifierTable::IdentifierTable(const LangOptions &LangOpts,
 // Language Keyword Implementation
 //===----------------------------------------------------------------------===//
 
-// Constants for TokenKinds.def
```
Sure, it used to be on the anonymous namespace for TokenKind and KeywordStatus, I've restored it on TokenKind only because KeywordStatus is only used by the code that uses TokenKinds.def not TokenKinds.def itself
